rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,634 @@
+
+    >>> a = tf.Module()
+    >>> b = tf.Module()
+    >>> c = tf.Module()
+    >>> a.b = b
+    >>> b.c = c
+    >>> assert list(a.submodules) == [b, c]
+    >>> assert list(b.submodules) == [c]
+    >>> assert list(c.submodules) == []
+
+    Returns:
+      A generator over all submodules.
+    """
+
+  @property
+  def owned_submodules(self):
+    """Returns an iterator of immediate child modules.
+
+    Child modules are modules which are found as properties of the current
+    module.
+
+    >>> a = tf.Module()
+    >>> b = tf.Module()
+    >>> c = tf.Module()
+    >>> a.b = b
+    >>> b.c = c
+    >>> assert list(a.owned_submodules) == [b]
+    >>> assert list(b.owned_submodules) == [c]
+    >>> assert list(c.owned_submodules) == []
+
+    Returns:
+      A generator over all child modules.
+    """
+
+class ModuleMetaclass(type):
+  def __new__(mcs, name, bases, clsdict):
+    """Modifies module subclasses when they are defined.
+
+    Wraps all methods and properties in clsdict such that enter the modules
+    name scope before evaluation.
+
+    >>> class MyModule(tf.Module):
+    ...   def __call__(self, x):
+    ...     if not hasattr(self, 'w'):
+    ...       self.w = tf.Variable(tf.zeros([x.shape[1], 64]), name='w')
+    ...     return tf.matmul(x, w)
+
+    Without any user intervention __call__ will enter the modules name scope,
+    allowing variables and submodules to be created. In some cases this is not
+    desirable, so there exists an opt out annotation(`@no_auto_name_scope`)
+    which instructs the metaclass to not wrap these functions.
+    """
+
+  def __call__(cls, *args, **kwargs):
+    """Creates new module instances.
+
+    To allow other modules/variables to be created in the constructor of a
+    module (and be named appropriately) we enter the module name scope inside
+    the base (Module) constructor. We close the name scope in the metaclass
+    before returning the new instance.
+
+    >>> class MyModule(tf.Module):
+    ...   def __init__(self):
+    ...     super(MyModule, self).__init__()
+    ...     # ^ Enters the module name scope but does not exit.
+    ...
+    ...     self.v = tf.Variable(1., name='v')
+    ...     # ^ Will be named "my_module/v"
+    ...
+    ...     self.m = MyOtherModule()
+    ...     # ^ Will be named "my_module/my_other_module"
+
+    >>> m = MyModule()
+    >>> # ^ Name scope will be closed thanks to metaclass.
+    """
+```
+
+### Naming
+
+Having human readable names for ops and variables is essential for debugging.
+Each `tf.Module` will have a name scope which is determined in the constructor.
+By default this will be a `lower_camel_case` version of the `ClassName`:
+
+```python
+class MyModule(tf.Module):
+  def __init__(self, name=None):
+    super(MyModule, self).__init__(name=name)
+    self.v = tf.Variable(1., name='v')
+
+>>> m = MyModule()
+>>> m.name
+'my_module'
+>>> m.name_scope.name
+'my_module/'
+>>> m.v.name
+'my_module/v:0'
```

Comment:
Are the variable names (with scope) attached upon the variable creation? An alternative might be the pytorch way that a variable does not know about its (parent's) scope, and the full scoped name is only constructed on the fly depending on which parent module are we dealing with (because a module can be re-used in different parent modules, and they might even be constructed externally and pass in via the constructor of a custom module).

I'm thinking about the following pseudocode example:
```python
class FunnyAutoEncoder(tf.Module):
  def __init__(self, encoder, decoder, name=None):
    super(...)
    self.encoder = encoder
    self.decoder = decoder

  def __call__(self, x):
    z = self.encoder(x)
    z2 = tf.funny_op(z)
    return self.decoder(z2)


encoder = MLP(...)
decoder = MLP(...)
ae = FunnyAutoEncoder(encoder, decoder)
```

then `ae.encoder.layerX.w.name` will probably not show anything related to being under an auto encoder? Instead, it will be just with a top-level `'mlp/'` scope?

I think a major scenario of having full scoped names available is to iterate through all the variables in a module, and do some inspection / processing to the variable according to the names. In this case, it might be more useful to have the (scoped) names reflect what role each variable is playing in the module.

However, I guess one drawback might be to detect potential duplication. For example, if one does this
```python
encoder = ResidualBlock(...)
decoder = encoder
ae = FunnyAutoEncoder(encoder, decoder)
```
then the same variables might be listed twice under different (scoped) names. But I guess this situation is rather rare, and it is probably to explicitly define a `FunnyAutoEncoderSharedEncoderDecoder` that has only one `encoder_and_decoder` field.

Reply:
So currently the proposal is that name scopes are determined when the module is constructed and from that point it is fixed per module. For name based filtering you can do something like this:

```python
encoder = MLP(..., name='encoder')
decoder = MLP(..., name='decoder')
ae = FunnyAutoEncoder(encoder, decoder)

encoder_vars = encoder.variables
# .. or ..
encoder_vars = [v for v in ae.variables if v.name.startswith('encoder/')]
```

It's worth pointing out that you can also enter a name scope and build modules inside it if you want to group a set of modules that way:

```python
with tf.name_scope('encoder'):
  encoder = MLP(...)  # name_scope will be `encoder/mlp/`
decoder = MLP(..., name='decoder')  # name_scope := `decoder/`
ae = FunnyAutoEncoder(encoder, decoder)  # name_scope := `funny_auto_encoder/`
```

Is that flexible enough for your use cases? We've had a similar pattern in Sonnet and I've not heard any complaints from users about it, but perhaps there's something I'm missing!

Reply:
@tomhennigan Thanks! I'm not sure if it is because I'm too used to the other way of thinking. Let me state the use cases in my mind and we can see if it is easy to do in the current proposal. I think the main inflexibility of the current approach is that in order to have a meaningful namescope hierarchy, we can only construct the model from top-down according to the hierarchy. It will be difficult, for example, to construct some part of the component (e.g. the `encoder` from the example above) without knowing *beforehand* what that component is going to be used for (i.e. in your example, you explicitly pass the `name='encoder'` argument, indicating you know that it is going to be used as an encoder).

This problem might be more pronounced when one want to write a generic library of commonly used (sub)nets. For example, in library1 (say Sonnet2), I have a `Module` that defines a default ResNet implementation. In library2, which is a general RL training framework, that allow one to pass in a `Module` to be used for each specific component (e.g. the Vision part of the agent). Now as the user, if I want to use the ResNet in the vision part of the agent, I might need to use some hack to figure out what proper `name` to pass into when constructing the ResNet to make it consistent with the naming hierarchy in the agent model. Moreover, if the agent is grouping all its vision part code under the name scope `agent/vision`, then I cannot pass `agent/vision` as `name=` parameter to `ResNet` because it is not a proper python variable name.

My guess that this is not a problem in Sonnet(v1) is that Sonnet(v1) is still graph based library with explicit `build()` function. In other words, when you create the Python object, the underlying graph building (therefore the naming and scoping) has not been called yet. Therefore, one can also change the name scoping *when* the build is actually being called. Here is some pseudo-code of my understanding of what might happen in Sonnet V1
```python
class Agent(...):
  def __init__(self, vision_net, ...):
    self.vision_net = vision_net
    # ....

  def build(self, ...):
    with tf.name_scope(self.name + '/vision'):
      self.vision_net.build(...)
```
On the other hand, with the current proposal, users could (and it seems recommended) create variable in the python object `__init__` function, which will attach name to the variable upon creation. Therefore, one need to know the exact name of the parent *hierarchy* upon creating that python object. Once the `Module` object is created, the parent module can no longer re-wire its name scope as in the case of Sonnet V1. Is my understanding correct here?

Reply:
I agree with @pluskid that dependency injection should at least be possible. My preference is for DI to be the recommended way to build up modules, but I'd understand if others disagreed.

Reply:
Thanks for the IRL discussions @pluskid, @brilee, @superbobry and @chr1sj0nes.

To summarize our discussions there are many cases where there are multiple paths to weights in a model (e.g. `model.encoder.w` might be the same instance as `model.decoder.w`) and this is somewhat confusing since in TensorFlow variables have a single, immutable name (e.g. `w`'s name would be `encoder/w:0` in both cases if encoder created it).

Perhaps the "right" solution is to remove the name from variables in Python, and instead use Python variable names to "name" variables. Sadly this is not in scope for TensorFlow 2, and we agreed that the current strategy of variable naming at creation time was OK for simple cases and better than not naming.

We discussed that the python variables or object structure can be used instead to associate a path which each parameter and in the above case you would have two paths to `w`. We agreed that it would be convenient to have some easy way to traverse the structure of arbitrary `Module`s in a way that have some simple protocol (e.g. somewhere inside the model there is a property called `encoder` and somewhere inside the model there is a property called `decoder`) to filter these paths.

In tensorflow/tensorflow@1d5fd9219 I added `with_path` to Module's `_flatten` method. This allows you to do exactly that. For example you can find paths in some arbitrary model to all `tf.Variable`s inside a (possibly deeply nested) `Module` in a property called `encoder`:

```python
>>> class MyModule(tf.Module):
...   @property
...   def state_dict(self):
...     return dict(self._flatten(
...         predicate=lambda v: isinstance(v, tf.Variable), with_path=True))

>>> mod = MyModule()
>>> mod.encoder = Encoder()
>>> mod.decoder = mod.encoder

>>> mod.state_dict
{('encoder', 'w'), <tf.Variable: ...>,
 ('decoder', 'w'), <tf.Variable: ...>}

>>> {k: v for k, v in mod.state_dict.items() if 'encoder' in k}
{('encoder', 'w'), <tf.Variable: ...>}
```

I hope that helps :smile:. h/t @gehring who was asking about how to determine "depth" in another comment. I think this `with_path` option to `_flatten` (or the `recursive` flag) would help solve that too.